clik/external/hal_cpu/source/cpu_hal.cpp

```diff
@@ -147,10 +174,40 @@ void cpu_barrier::wait(int num_threads) {
   }
 }
 
+#if defined(HAL_CPU_WI_MODE)
 void cpu_hal::kernel_entry(exec_state *exec) {
   direct_kernel_fn kernel = (direct_kernel_fn)exec->kernel_entry;
   kernel((void *)exec->packed_args, exec);
 }
+#else
+void cpu_hal::kernel_entry(exec_state *exec) {
+  direct_kernel_fn kernel = (direct_kernel_fn)exec->kernel_entry;
+
+  // Simple split across x axis by dividing the num groups by the threads and
+  // rounding up. Some threads may get zero.
+  uint32_t num_groups_per_thread =
+      (exec->wg.num_groups[0] + exec->num_threads - 1) / exec->num_threads;
+
+  // Calculate the start x per thread
+  uint32_t group_x_start = num_groups_per_thread * exec->thread_id;
+
+  // Calculate the minumum of the next thread's group start and the max groups,
+  // so we know when to stop.
+  uint32_t group_x_next = std::min(group_x_start + num_groups_per_thread,
+                                   (uint32_t)(exec->wg.num_groups[0]));
+
+  for (uint32_t wg_z = 0; wg_z < exec->wg.num_groups[2]; wg_z++) {
+    exec->wg.group_id[2] = wg_z;
+    for (uint32_t wg_y = 0; wg_y < exec->wg.num_groups[1]; wg_y++) {
+      exec->wg.group_id[1] = wg_y;
+      for (uint32_t wg_x = group_x_start; wg_x < group_x_next; wg_x++) {
+        exec->wg.group_id[0] = wg_x;
+        kernel((void *)exec->packed_args, exec);
```

Comment:
I know this was already there for the WI mode, but what is the cast doing here?

Reply:
I've removed it.